20220831-add-type-restrictions-to-json-syntax.md

```diff
@@ -0,0 +1,570 @@
+
+#### Improve the Developer Experience
+
+At the moment, our authorization model does not allow users to indicate what the type they expect on a relation to be, leaving it ambiguous and potentially a cause of errors, especially when someone is reading the authorization model later without much context and where the original authors are not available. It is hard for someone to grok from first glance that a repository owner should be an organization or a user. Adding type restrictions allows developers to have an easier time in interpreting the model.
+
+### What is the expected outcome?
+
+- Using the extra type information we get from this change, we will be able to more easily traverse the graph in reverse, thus allowing us to provide a solution for reverse expansion and optimizing the ListObjects endpoint.
+- Users reading an authorization model will be able to better interpret its meaning (for example by understanding that repository owners must be either users or group members).
+
+## What it is?
+
+### Adding Type Restrictions
+
+This RFC proposes an update to the OpenFGA JSON syntax requiring users to indicate all the types of users that could be **directly related** to an object of a certain type through a particular relation.
+
+For example, our current syntax allows expressing the following:
+
+- documents have parents
+- repositories have owners
+
+But cannot express restrictions such as:
+
+- parents of a document have to be objects of type folder
+- owners of a repository have to be of type user or a userset of group members
+
+We introduce a `metadata` entry in the type definition would look like:
+
+```typescript
+type Metadata = {
+  relations?: Record<string, RelationMetadata>;
+};
+
+type RelationMetadata = {
+  allow_public?: Boolean;
+  directly_related_user_types?: DirectlyRelatedUserType[];
+};
+
+type DirectlyRelatedUserType = {
+  type: string;
+  relation?: string;
+};
+```
+
+The following is an example of the changes we are proposing, an explanation of the additions will follow:
+
+```json
+{ "type_definitions": [
+  { "type": "user", "relations": {} },
+  { "type": "employee", "relations": {} },
+  { "type": "group",
+    "relations": {
+      "parent": { "this": {} },
+      "member": { "this": {} },
+      "guest": { "this": {} },
+    },
+    "metadata": {
+      "relations": {
+        "parent": { "directly_related_user_types": [{ "type": "group" }] },
+        "member": { "directly_related_user_types": [{ "type": "employee" }, { "type": "group", "relation": "member" }] },
+        "guest": { "directly_related_user_types": [{ "type": "user" }, { "type": "employee" }, { "type": "group", "relation": "member" }], "allow_public": true }
+      }
+    }
+  }
+] }
+```
+
+A `directly_related_user_types` array to each relation to indicate what types of users can be directly related to the relation. It will be an array of objects, each object must have a type and an optional relation.
```

Comment:
I think we should revisit this name a bit. This restriction should restrict by object relation, not just object type. So calling it `directly_related_user_types` is a bit misleading IMO.

For example, we could have `user` or `group#member` as a restriction. So what we're technically restricting on is a relation of a particular object type. I like to think of it more like we're restricting `user#...` or `group#member`, but since we do not represent ellipses in anyway, we just say `user` or `group#member`